File: crates/hologram-ffi/interfaces/typescript/examples/executor_management.ts
/**
 * Executor Management Example
 * 
 * This example demonstrates comprehensive executor management including creation,
 * buffer allocation, phase management, resonance tracking, and topology operations.
 */

import * as hg from '../src/index';

console.log('🚀 Hologram FFI - Executor Management Example');
console.log('=============================================');

// Create a new executor
console.log('\n🏗️ Creating Executor:');
const executorHandle = hg.newExecutor();
console.log(`Created executor with handle: ${executorHandle}`);

// Create executor with custom backend
console.log('\n🔧 Creating Executor with Custom Backend:');
const customExecutorHandle = hg.executorWithBackend('cpu');
console.log(`Created custom executor with handle: ${customExecutorHandle}`);

// Get executor phase
console.log('\n📊 Executor Phase Management:');
const phase = hg.executorPhase(executorHandle);
console.log(`Current executor phase: ${phase}`);

// Advance executor phase
hg.executorAdvancePhase(executorHandle, 5);
const newPhase = hg.executorPhase(executorHandle);
console.log(`Phase after advancing by 5: ${newPhase}`);

// Allocate buffers
console.log('\n💾 Buffer Allocation:');
const bufferSize = 1024;
const bufferHandle = hg.executorAllocateBuffer(executorHandle, bufferSize);
console.log(`Allocated buffer with handle: ${bufferHandle}, size: ${bufferSize}`);

// Try to allocate boundary buffer
console.log('\n🔲 Boundary Buffer Allocation:');
try {
  const boundaryBufferHandle = hg.executorAllocateBoundaryBuffer(executorHandle, 0, 32, 32);
  console.log(`Allocated boundary buffer with handle: ${boundaryBufferHandle}`);
  
  // Check buffer properties
  const isBoundary = hg.bufferIsBoundary(boundaryBufferHandle);
  const isLinear = hg.bufferIsLinear(boundaryBufferHandle);
  const elementSize = hg.bufferElementSize(boundaryBufferHandle);
  const sizeBytes = hg.bufferSizeBytes(boundaryBufferHandle);
  
  console.log(`Boundary buffer properties:`);
  console.log(`  Is boundary: ${isBoundary}`);
  console.log(`  Is linear: ${isLinear}`);
  console.log(`  Element size: ${elementSize} bytes`);
  console.log(`  Total size: ${sizeBytes} bytes`);
  
  // Clean up boundary buffer
  hg.bufferCleanup(boundaryBufferHandle);
  console.log('Boundary buffer cleaned up');
} catch (error) {
  console.log(`Boundary buffer allocation failed: ${error}`);
}

// Buffer operations
console.log('\n🔄 Buffer Operations:');
const length = hg.bufferLength(bufferHandle);
console.log(`Buffer length: ${length}`);

// Fill buffer with test data
hg.bufferFill(bufferHandle, 1.5, bufferSize);
console.log(`Filled buffer with value 1.5`);

// Get buffer data
const bufferData = hg.bufferToVec(bufferHandle);
console.log(`Buffer data (first 10 elements): ${bufferData.substring(0, 100)}...`);

// Check buffer properties
const isEmpty = hg.bufferIsEmpty(bufferHandle);
const isLinear = hg.bufferIsLinear(bufferHandle);
const isBoundary = hg.bufferIsBoundary(bufferHandle);
const elementSize = hg.bufferElementSize(bufferHandle);
const sizeBytes = hg.bufferSizeBytes(bufferHandle);
const backendHandle = hg.bufferBackendHandle(bufferHandle);
const topology = hg.bufferTopology(bufferHandle);
const pool = hg.bufferPool(bufferHandle);

console.log('\n📋 Buffer Properties:');
console.log(`  Is empty: ${isEmpty}`);
console.log(`  Is linear: ${isLinear}`);
console.log(`  Is boundary: ${isBoundary}`);
console.log(`  Element size: ${elementSize} bytes`);
console.log(`  Total size: ${sizeBytes} bytes`);
console.log(`  Backend handle: ${backendHandle}`);
console.log(`  Topology: ${topology}`);
console.log(`  Pool: ${pool}`);

// Copy data to buffer
console.log('\n📋 Copying Data to Buffer:');
const testData = JSON.stringify([1.0, 2.0, 3.0, 4.0, 5.0]);
hg.bufferCopyFromSlice(bufferHandle, testData);
console.log(`Copied test data to buffer`);

// Resonance operations
console.log('\n🎵 Resonance Operations:');
for (let i = 0; i < 5; i++) {
  const resonance = hg.executorResonanceAt(executorHandle, i);
  console.log(`Resonance at class ${i}: ${resonance.toFixed(4)}`);
}

// Get resonance snapshot
const resonanceSnapshot = hg.executorResonanceSnapshot(executorHandle);
console.log(`Resonance snapshot: ${resonanceSnapshot}`);

// Mirror operations
console.log('\n🪞 Mirror Operations:');
for (let i = 0; i < 5; i++) {
  const mirror = hg.executorMirror(executorHandle, i);
  console.log(`Mirror of class ${i}: ${mirror}`);
}

// Neighbor operations
console.log('\n👥 Neighbor Operations:');
for (let i = 0; i < 5; i++) {
  const neighbors = hg.executorNeighbors(executorHandle, i);
  console.log(`Neighbors of class ${i}: ${neighbors}`);
}

// Cleanup
console.log('\n🧹 Cleanup:');
hg.bufferCleanup(bufferHandle);
console.log('Buffer cleaned up');

hg.executorCleanup(executorHandle);
console.log('Executor cleaned up');

hg.executorCleanup(customExecutorHandle);
console.log('Custom executor cleaned up');

console.log('\n✅ Executor management example completed successfully!');

File: crates/hologram-ffi/interfaces/python/examples/executor_management.py
#!/usr/bin/env python3
"""
Executor Management Example

This example demonstrates comprehensive executor management including:
- Creating executors with different backends
- Buffer allocation (linear and boundary)
- Phase management
- Resonance tracking
- Topology operations (mirrors, neighbors)
- Proper cleanup
"""

import hologram_ffi as hg
import json

def main():
    print("=" * 60)
    print("Hologram FFI - Executor Management Example")
    print("=" * 60)
    
    # Create executor
    print("\n🏗️ Creating Executor:")
    try:
        executor_handle = hg.new_executor()
        print(f"✅ Created executor with handle: {executor_handle}")
    except Exception as e:
        print(f"❌ Failed to create executor: {e}")
        return
    
    # Test executor with custom backend
    print("\n🔧 Testing Custom Backend:")
    try:
        custom_executor_handle = hg.executor_with_backend("cpu")
        print(f"✅ Created executor with custom backend: {custom_executor_handle}")
        hg.executor_cleanup(custom_executor_handle)
        print("✅ Cleaned up custom executor")
    except Exception as e:
        print(f"❌ Custom backend test failed: {e}")
    
    # Test buffer allocation
    print("\n📦 Buffer Allocation:")
    try:
        # Linear buffer
        linear_buffer_handle = hg.executor_allocate_buffer(executor_handle, 1024)
        print(f"✅ Allocated linear buffer: {linear_buffer_handle}")
        
        # Boundary buffer (may fail on some systems)
        try:
            boundary_buffer_handle = hg.executor_allocate_boundary_buffer(executor_handle, 0, 32, 32)
            print(f"✅ Allocated boundary buffer: {boundary_buffer_handle}")
        except Exception as e:
            print(f"⚠️ Boundary buffer allocation failed (expected on some systems): {e}")
            boundary_buffer_handle = None
        
    except Exception as e:
        print(f"❌ Buffer allocation failed: {e}")
        boundary_buffer_handle = None
    
    # Test phase management
    print("\n⏰ Phase Management:")
    try:
        current_phase = hg.executor_phase(executor_handle)
        print(f"Current phase: {current_phase}")
        
        hg.executor_advance_phase(executor_handle, 5)
        new_phase = hg.executor_phase(executor_handle)
        print(f"Phase after advancing by 5: {new_phase}")
        
    except Exception as e:
        print(f"❌ Phase management failed: {e}")
    
    # Test resonance tracking
    print("\n🌊 Resonance Tracking:")
    try:
        # Get resonance for specific class
        resonance_0 = hg.executor_resonance_at(executor_handle, 0)
        resonance_42 = hg.executor_resonance_at(executor_handle, 42)
        print(f"Resonance at class 0: {resonance_0}")
        print(f"Resonance at class 42: {resonance_42}")
        
        # Get full resonance snapshot
        snapshot = hg.executor_resonance_snapshot(executor_handle)
        snapshot_data = json.loads(snapshot)
        print(f"Resonance snapshot: {len(snapshot_data)} classes")
        
    except Exception as e:
        print(f"❌ Resonance tracking failed: {e}")
    
    # Test topology operations
    print("\n🔗 Topology Operations:")
    try:
        # Mirror operations
        mirror_0 = hg.executor_mirror(executor_handle, 0)
        mirror_42 = hg.executor_mirror(executor_handle, 42)
        print(f"Mirror of class 0: {mirror_0}")
        print(f"Mirror of class 42: {mirror_42}")
        
        # Neighbor operations
        neighbors_0 = hg.executor_neighbors(executor_handle, 0)
        neighbors_data = json.loads(neighbors_0)
        print(f"Neighbors of class 0: {neighbors_data}")
        
    except Exception as e:
        print(f"❌ Topology operations failed: {e}")
    
    # Cleanup
    print("\n🧹 Cleanup:")
    try:
        hg.buffer_cleanup(linear_buffer_handle)
        if boundary_buffer_handle is not None:
            hg.buffer_cleanup(boundary_buffer_handle)
        hg.executor_cleanup(executor_handle)
        print("✅ All resources cleaned up successfully")
    except Exception as e:
        print(f"❌ Cleanup failed: {e}")
    
    print("\n✅ Executor management example completed!")

if __name__ == "__main__":
    main()
